### DISCUSSION TITLE
EMG channel is not detected during the sleep stage classification

### DISCUSSION THREAD

**Discussion by IstS-KV** [raphaelvallat/yasa#178] [Q&A] [answered]
Hi
While working on sleep stage classification, I have noticed that algorithm doesn't extract features for EMG channel.

My .edf files contain recordings from multiple modalities. I extract only EEG channels of interest, EOG and EMG. However, due to the naming EMG channel is by default recognised as EEG. That is why I change its type with the code line:  data.set_channel_types({'EMG1-EMG2': 'emg'}).
After the initialization of the classifier, I have printed out the features that were calculated by the algo. The dataframe contains only features for EEG and EOG data.

Here is the part of the code:

     ch_list = ['Fp2-P4', 'F4-C4', 'C4-P4', 'C4-A1', 'ROC-LOC', 'EMG1-EMG2']
    
    # read the file
    data = mne.io.read_raw_edf(file, eog = ['ROC-LOC'], include = ch_list, preload=True, verbose=False) 
    data.set_channel_types({'EMG1-EMG2': 'emg'})
    
    # output infprmation about the recordings
    print('')
    print(data.info['sfreq'])
    print(data.info['ch_names'])
    print(data.get_channel_types())
    print('')
    # input('wait...')

    # filtering EEG signal
    data = data.filter(0.1, 35)

    # Initialize the sleep stage algo
    sls = yasa.SleepStaging(data, eeg_name='C4-A1' , eog_name ='ROC-LOC' , emg_name='EMG1-EMG2')

    # idouble check the extracted features
    print(sls.get_features().columns.values.tolist())
    input('Press enter to continue')
    # Make predictions
    y_pred = sls.predict()

During printing data.get_channel_types(), I can see that EMG1-EMG2 now is recognised as emg. However, I was thinking that the problem occurs because I had changed the channel type manually. If so, I need to change the name of the EMG channel in .edf file before I read it, so it would be recognised automatically by the algo. 
I would like to know if my assumption is correct and if there any other solution to solve it?

Best,
Khrystyna

**Comment by remrama** [comment, marked as answer]
Hi, here's a few thoughts.

YASA doesn't really care about the MNE channel types. When you specify the channels during the initialization of a `SleepStaging` class (e.g., `yasa.SleepStaging(raw, emg_name="EMG1-EMG2")`), that will override whatever the MNE channel types are. So if you are not getting EMG features, the issue probably lies somewhere else.

I used the YASA sample data file from the [Quickstart page](https://raphaelvallat.com/yasa/build/html/quickstart.html) (or direct link [here](https://drive.google.com/uc?id=13YFQrvAGvcz77Qm8kFXu1nyRKhTti3VB)) to show 2 things: (1) YASA doesn't care about MNE channel types, and (2) YASA is generating EMG features, at least in my setup.

```python
import mne
import yasa

# Load and preprocess sample data with 3 channels
channels = ["C3-A2", "EMG1-EMG2", "ROC-A1"]
raw = mne.io.read_raw_edf("yasa_example_night_young.edf", include=channels, preload=True, verbose=False)
raw.resample(100)
raw.filter(0.3, 45, verbose=False)

# Get a list of default MNE channel types
ch_types = raw.get_channel_types()

# Run YASA sleep staging with default MNE channel types
sls = yasa.SleepStaging(raw, eeg_name="C3-A2", emg_name="EMG1-EMG2", eog_name="ROC-A1")
features = sls.get_features()
emg_columns = features.filter(like="emg").columns.tolist()
n_emg_columns = len(emg_columns)

# Modify the MNE channel types
raw.set_channel_types({"EMG1-EMG2": "emg", "ROC-A1": "eog"})
modified_ch_types = raw.get_channel_types()

# Rerun YASA sleep staging with modified MNE channel types
modified_sls = yasa.SleepStaging(raw, eeg_name="C3-A2", eog_name="ROC-A1", emg_name="EMG1-EMG2")
modified_features = modified_sls.get_features()
modified_emg_columns = modified_features.filter(like="emg").columns.tolist()
modified_n_emg_columns = len(modified_emg_columns)

# Check if the results are identical no matter what the MNE channel types are
equal_features = features.equals(modified_features)

# Print results
print("====== Sleep Staging with Default MNE Channel Types ".ljust(75, "="))
print("Default channel types:", ch_types)
print("Number of YASA EMG features:", n_emg_columns)
print("First 3 YASA EMG columns:", emg_columns[:3])
print("====== Sleep Staging with Modified MNE Channel Types ".ljust(75, "="))
print("Modified channel types:", modified_ch_types)
print("Number of YASA EMG features:", modified_n_emg_columns)
print("First 3 YASA EMG columns:", modified_emg_columns[:3])
print("====== Verifying that YASA is Agnostic to MNE Channel Types ".ljust(75, "="))
print("Are the two features DataFrames equal?", equal_features)
```

My output from that is:

```
====== Sleep Staging with Default MNE Channel Types =======================
Default channel types: ['eeg', 'eeg', 'eeg']
Number of YASA EMG features: 33
First 3 YASA EMG columns: ['emg_abspow', 'emg_abspow_c7min_norm', 'emg_abspow_p2min_norm']
====== Sleep Staging with Modified MNE Channel Types ======================
Modified channel types: ['eog', 'eeg', 'emg']
Number of YASA EMG features: 33
First 3 YASA EMG columns: ['emg_abspow', 'emg_abspow_c7min_norm', 'emg_abspow_p2min_norm']
====== Verifying that YASA is Agnostic to MNE Channel Types ===============
Are the two features DataFrames equal? True
```

Could you take the following steps?

1. Double-check your results after removing the channel type modifications? This might simplify the code to help isolate the error.
2. Rerun your code using a new EDF file? Try it first with the same sample YASA file I used, and if that works, move to another file from your study from a different participant. If it doesn't work with the YASA file, that suggests it's something to do with your YASA installation.
3. If it doesn't work for you on the YASA sample file, uninstall and reinstall YASA and make sure you have the latest version with ``pip install --upgrade --force-reinstall yasa``.
4. If it doesn't work on the sample file after an upgrade/reinstallation, let us know some details about the Python environment you're using (YASA version, MNE version, operating system, etc.).
5. If it works on your YASA file but not any of your own EDF files, hopefully you're able to provide an anonymized sample file here and we can try to diagnose the problem with it.
